divide_and_conquer/convex_hull.py

```diff
@@ -0,0 +1,437 @@
+from __future__ import print_function, absolute_import, division
+
+from numbers import Number
+"""
+The convex hull problem is problem of finding all the vertices of convex polygon, P of 
+a set of points in a plane such that all the points are either on the vertices of P or
+inside P. TH convex hull problem has several applications in geometrical problems, 
+computer graphics and game development. 
+
+Two algorithms have been implemented for the convex hull problem here. 
+1. A brute-force algorithm which runs in O(n^3)
+2. A divide-and-conquer algorithm which runs in O(n^3)
+
+There are other several other algorithms for the convex hull problem
+which have not been implemented here, yet. 
+
+"""
+
+
+class Point:
+    """
+    Defines a 2-d point for use by all convex-hull algorithms.
+
+    Parameters
+    ----------
+    x: an int or a float, the x-coordinate of the 2-d point
+    y: an int or a float, the y-coordinate of the 2-d point
+
+    Examples
+    --------
+    >>> Point(1, 2)
+    (1, 2)
+    >>> Point("1", "2")
+    (1.0, 2.0)
+    >>> Point(1, 2) > Point(0, 1)
+    True
+    >>> Point(1, 1) == Point(1, 1)
+    True
+    >>> Point(-0.5, 1) == Point(0.5, 1)
+    False
+    >>> Point("pi", "e")
+    Traceback (most recent call last):
+        ...
+    ValueError: x and y must be both numeric types but got <class 'str'>, <class 'str'> instead
+     """
+
+    def __init__(self, x, y):
+        if not (isinstance(x, Number) and isinstance(y, Number)):
+            try:
+                x, y = float(x), float(y)
+            except ValueError as e:
+                e.args = ("x and y must be both numeric types "
+                          "but got {}, {} instead".format(str(type(x)), str(type(y))), )
+                raise
+
+        self.x = x
+        self.y = y
+
+    def __eq__(self, other):
+        return self.x == other.x and self.y == other.y
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __gt__(self, other):
+        if self.x > other.x:
+            return True
+        elif self.x == other.x:
+            return self.y > other.y
+        return False
+
+    def __lt__(self, other):
+        return not self > other
+
+    def __ge__(self, other):
+        if self.x > other.x:
+            return True
+        elif self.x == other.x:
+            return self.y >= other.y
+        return False
+
+    def __le__(self, other):
+        if self.x < other.x:
+            return True
+        elif self.x == other.x:
+            return self.y <= other.y
+        return False
+
+    def __repr__(self):
+        return "({}, {})".format(self.x,  self.y)
```

Comment:
Try this one with an f”string” and see if it is more concise.

Reply:
But f'strings' don't work with python 2 or?